# Initial data cleaning
Ondrej Bohdal

# Notes on what to do

## Already done
•	Step 0: Rearrange cols so that they match in all files.

•	Step 1: Create link between file name and search variables (Industry / Function / Years of Experience) in new columns

•	Step 2: Combine all files together into one gigantic spreadsheet

•	Step 3: standardize location names

◦	use find and replace for following terms: *waterloo* *kitchener* *guelph*, replace all with Waterloo

▪	issues: might affect firm names, other details, might have to do individual location columns - solved by dealing only with location columns

▪	can’t do this for Cambridge, since that’ll take out the better Cambridges in the US and UK - there are very little of Cambridge instances (if any, but definitely less than 200), so it is not a problem.

•	Step 4: standardize Blackberry

◦	Use find and replace for following terms *Research in Motion* *RIM* *Blackberry*, replace with RIM

•	Step 5 - clean employment date columns

◦	delineate column based on '('

◦	There’s currently a problem in the scrapper so that about 1% of the time this is a very long cell with lots of extraneous info. Right now have to remove them manually else they’ll mess up the text to column function in excel

◦	create new column after employment date column, split cell based on ‘(‘, delate data in new column

◦	next step: need to make this computer readable. Ideas: change all ‘Present’ to 2016, remove all text from columns, allowing 
us to get years they were employed.

•	Step 6: Create schema of education (e.g. BA / Bachelors / BSc / B.Tech) and look for way to standardise

•	Goal 1: Find way to systematically identify people who have worked in Waterloo AND worked for RIM

•	Goal 2: Identify point where they stopped working for RIM pre- and post- 2008 and extract their subsequent jobs and locations

## To do
Everything done currently, so just explore the data further.

For the beginning import all necesssary libraries.

In [1]:
import pandas as pd
import numpy as np
import os

# Processing of the files
Manually convert the files to csv - the number of files is not too big, and finding a solution to do it automatically would take more time than just doing it manually.

Open each of the files in Excel, and save it as csv with the original name. If there are any problems just continue by accepting it.

Let's get a list of all the csv files.

In [2]:
print(os.listdir('LinkedInDataCsv'))

['10.csv', '11.csv', '12.csv', '13.csv', '15.csv', '16a.csv', '16b.csv', '17.csv', '18.csv', '19.csv', '21.csv', '22.csv', '23.csv', '24.csv', '25.csv', '27a.csv', '27b.csv', '27c.csv', '28.csv', '28b.csv', '29.csv', '3.csv', '30.csv', '31.csv', '4.csv', '9.csv']


We save all the names of the files so that we can process them.

In [3]:
files_to_process = ['10.csv', '11.csv', '12.csv', '13.csv', '15.csv', '16a.csv', '16b.csv', '17.csv', '18.csv', '19.csv', '21.csv', '22.csv', '23.csv', '24.csv', '25.csv', '27a.csv', '27b.csv', '27c.csv', '28.csv', '28b.csv', '29.csv', '3.csv', '30.csv', '31.csv', '4.csv', '9.csv']

Save the dataframes into a list, and then we will merge them. During this we add a new column with the name of the original file name - we give there the name with .xls (so it is e.g. 10.xls).

In [4]:
data_list = []
for file in files_to_process:
    data_frame_local = pd.read_csv('LinkedInDataCsv/' + file, encoding = "ISO-8859-1")
    data_frame_local['File name'] = file[:-4] + '.xls'
    data_list.append(data_frame_local)

Let's concatenate all of the data frames into one gigantic data frame.

In [5]:
data = pd.concat(data_list, ignore_index=True)

The order of column names is now sorted alphabetically, but we would like them to be sorted by their meaning so that it is similar to the files.

In [6]:
important = ['Source Url', 'Name', 'Location', 'Industry', 'University', 'Field Of Study', 'Number Of Connections', 'Job Title', 'Employer', 'Employment Date', 'Location 2', 'Job Title 2', 'Employer 2', 'Employment Date 2', 'Location 3', 'Job Title 3', 'Employer 3', 'Employment Date 3', 'Location 4', 'Job Title 4', 'Employer 4', 'Employment Date 4', 'Location 5', 'Job Title 5', 'Employer 5', 'Employment Date 5', 'Location 6', 'Job Title 6', 'Employer 6', 'Employment Date 6', 'Location 7', 'Job Title 7', 'Employer 7', 'Employment Date 7', 'Location 8', 'University 2', 'Degree', 'Degree Time', 'Degree Time 2', 'Location 9', 'Location 10', 'Job Title 8', 'Employer 8', 'Job Title 9', 'Employer 9', 'Job Title 10', 'Employer 10', 'Job Title 11', 'Employer 11', 'University 3', 'Degree 2', 'University 4', 'Degree 3', 'University 5', 'Degree 4', 'Employment Date 8', 'Employment Date 9', 'Employment Date 10', 'Location 11', 'Job Title 12', 'Employer 12', 'Employment Date 11', 'Location 12', 'Job Title 13', 'Employer 13', 'Employment Date 12', 'Location 13', 'Job Title 14', 'Employer 14', 'Job Title 15', 'Employer 15', 'Employment Date 13', 'Location 14', 'Job Title 16', 'Employer 16', 'Employment Date 14', 'Location 15', 'Job Title 17', 'Employer 17', 'Employment Date 15', 'Location 16', 'Field Of Study 2', 'Degree Time 3', 'Employment Date 16', 'Location 17', 'Job Title 18', 'Employer 18', 'Job Title 19', 'Employer 19', 'Employment Date 17', 'Location 18', 'Location 19', 'Field Of Study 3', 'Degree Time 4', 'Field Of Study 4', 'Degree Time 5', 'Degree Time 6', 'Job Title 20', 'Employer 20', 'Job Title 21', 'Employer 21', 'Job Title 22', 'Employer 22', 'Job Title 23', 'Employer 23', 'Job Title 24', 'Employer 24', 'Job Title 25', 'Employer 25', 'Job Title 26', 'Employer 26', 'Job Title 27', 'Employer 27', 'Job Title 28', 'Employer 28', 'Field Of Study 5', 'Degree 5', 'Employment Date 18', 'Employment Date 19', 'Location 20', 'Employment Date 20', 'Location 21', 'Employment Date 21', 'Location 22', 'Employment Date 22', 'Location 23', 'Employment Date 23', 'Location 24', 'Employment Date 24', 'Location 25', 'Job Title 29', 'Employer 29', 'Job Title 30', 'Employer 30', 'Job Title 31', 'Employer 31', 'Job Title 32', 'Employer 32', 'Job Title 33', 'Employer 33', 'Job Title 34', 'Employer 34', 'Job Title 35', 'Employer 35', 'Job Title 36', 'Employer 36', 'Employment Date 25', 'Location 26', 'Job Title 37', 'Employer 37', 'Job Title 38', 'Employer 38', 'Job Title 39', 'Employer 39', 'University 6', 'Degree 6', 'Degree 7', 'Degree Time 7', 'Degree 8', 'Degree Time 8', 'Degree 9', 'Degree Time 9', 'Degree 10', 'Degree Time 10', 'Degree 11', 'Degree Time 11', 'Field Of Study 6', 'Employment Date 26', 'Location 27', 'Employment Date 27', 'Location 28', 'Employment Date 28', 'Location 29', 'Employment Date 29', 'Location 30', 'Employment Date 30', 'Location 31', 'Employment Date 31', 'Location 32', 'Employment Date 32', 'Location 33', 'Employment Date 33', 'Location 34', 'Employment Date 34', 'Location 35', 'Employment Date 35', 'Location 36', 'Employment Date 36', 'Location 37', 'Employment Date 37', 'Location 38', 'Job Title 40', 'Employer 40', 'Employment Date 38', 'Location 39', 'Job Title 41', 'Employer 41', 'Employment Date 39', 'Location 40', 'Job Title 42', 'Employer 42', 'Employment Date 40', 'Location 41', 'Job Title 43', 'Employer 43', 'Employment Date 41', 'Location 42', 'University 7', 'Field Of Study 7', 'University 8', 'Field Of Study 8', 'University 9', 'Field Of Study 9', 'University 10']
order = important + [c for c in data.columns if c not in important]
data = data[order]

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5291 entries, 0 to 5290
Columns: 215 entries, Source Url to File name
dtypes: object(215)
memory usage: 8.7+ MB


In [8]:
data.describe()

Source Url             Name  \
count                                                5291             5291   
unique                                               5291             4790   
top     https://www.linkedin.com/profile/view?id=ADEAA...  LinkedIn Member   
freq                                                    1              307   

                    Location           Industry              University  \
count                   5291               5291                    4666   
unique                   494                 10                     883   
top     Toronto, Canada Area  Computer Software  University of Waterloo   
freq                     499               2440                     894   

          Field Of Study Number Of Connections          Job Title    Employer  \
count               4311                  5290               4981        4981   
unique              1057                   510               2554        2136   
top     Computer Science                  500+  Software Engineer  BlackBerry   
freq                1025                  1037                447         422   

                                   Employment Date    ...     \
count                                         4477    ...      
unique                                         837    ...      
top     September 2014 – Present (1 year 5 months)    ...      
freq                                            53    ...      

                                  Employment Date 41   Location 42  \
count                                              1             1   
unique                                             1             1   
top     July 1990 – December 1994 (4 years 6 months)  Osaka, Japan   
freq                                               1             1   

                 University 7 Field Of Study 7       University 8  \
count                      49                3                  8   
unique                     42                3                  7   
top     University of Toronto   Double Honours  McGill University   
freq                        4                1                  2   

           Field Of Study 8                                University 9  \
count                     2                                          15   
unique                    2                                          15   
top     Computer Programmer  University of Illinois at Urbana-Champaign   
freq                      1                                           1   

       Field Of Study 9           University 10 File name  
count                 2                       3      5291  
unique                2                       3        26  
top            Accounts  University of Waterloo    22.xls  
freq                  1                       1       700  

[4 rows x 215 columns]

In [9]:
print('These are the column names: ', list(data.columns))

These are the column names:  ['Source Url', 'Name', 'Location', 'Industry', 'University', 'Field Of Study', 'Number Of Connections', 'Job Title', 'Employer', 'Employment Date', 'Location 2', 'Job Title 2', 'Employer 2', 'Employment Date 2', 'Location 3', 'Job Title 3', 'Employer 3', 'Employment Date 3', 'Location 4', 'Job Title 4', 'Employer 4', 'Employment Date 4', 'Location 5', 'Job Title 5', 'Employer 5', 'Employment Date 5', 'Location 6', 'Job Title 6', 'Employer 6', 'Employment Date 6', 'Location 7', 'Job Title 7', 'Employer 7', 'Employment Date 7', 'Location 8', 'University 2', 'Degree', 'Degree Time', 'Degree Time 2', 'Location 9', 'Location 10', 'Job Title 8', 'Employer 8', 'Job Title 9', 'Employer 9', 'Job Title 10', 'Employer 10', 'Job Title 11', 'Employer 11', 'University 3', 'Degree 2', 'University 4', 'Degree 3', 'University 5', 'Degree 4', 'Employment Date 8', 'Employment Date 9', 'Employment Date 10', 'Location 11', 'Job Title 12', 'Employer 12', 'Employment Date 11', 'L

## Change the location names *waterloo*, *kitchener*, *guelph* to Waterloo.
This is not so trivial as the full names vary quite a lot, e.g. we may have Kitchener, Canada Area or also Kitchener, Ontario, Canada (and various alternatives). We will check if there is a substring of a location name including 'kitchener', 'guelp', 'waterloo' (after transforming to lower case).

We process only the location names so there is no risk of changing the names of any companies or anything unrelated.

Before we do it, let's do some exploration:

In [10]:
loc_names = ['Location'] + ['Location ' + str(x + 2) for x in range(41)]
nomatch = []
matches = []
for loc_name in loc_names:
    for i in range(len(data[loc_name])):
        if 'kitchener' in str(data[loc_name][i]).lower():
            matches.append(str(data[loc_name][i]))
        elif 'guelph' in str(data[loc_name][i]).lower():
            matches.append(str(data[loc_name][i]))
        elif 'waterloo' in str(data[loc_name][i]).lower():
            matches.append(str(data[loc_name][i]))
        else:
            nomatch.append(str(data[loc_name][i]))

Other common places include the following:
nan                                    197013,
Toronto, Canada Area                     1887,
Ottawa, Canada Area                      1757,
San Francisco Bay Area                    750,
Greater Seattle Area                      431,
Canada                                    380,
Dallas/Fort Worth Area                    304,
Vancouver, Canada Area                    293,
Ontario, Canada                           231,
Halifax, Canada Area                      206.

Note these are all places mentioned more than 200 times in any of the location columns. Waterloo (together with Guelph and Kitchener) is mentioned 4931 times, with Waterloo, Ontario exactly 1359 times.

In [11]:
print('Only Waterloo and related:')
print(pd.Series(matches).value_counts()[:5])
print('\nAll other locations:')
print(pd.Series(nomatch).value_counts()[:10])

Only Waterloo and related:
Waterloo, Ontario            1359
Kitchener, Canada Area        930
Waterloo, Ontario, Canada     785
Waterloo, ON                  572
Waterloo                      220
dtype: int64

All other locations:
nan                       197013
Toronto, Canada Area        1887
Ottawa, Canada Area         1757
San Francisco Bay Area       750
Greater Seattle Area         431
Canada                       380
Dallas/Fort Worth Area       304
Vancouver, Canada Area       293
Ontario, Canada              231
Halifax, Canada Area         206
dtype: int64


Let's transform the data now. The exact final name in all matches will be Waterloo, Ontario, Canada.

In [12]:
loc_names = ['Location'] + ['Location ' + str(x + 2) for x in range(41)]
for loc_name in loc_names:
    for i in range(len(data[loc_name])):
        if 'kitchener' in str(data[loc_name][i]).lower():
            data[loc_name][i] = 'Waterloo, Ontario, Canada'
        elif 'guelph' in str(data[loc_name][i]).lower():
            data[loc_name][i] = 'Waterloo, Ontario, Canada'
        elif 'waterloo' in str(data[loc_name][i]).lower():
            data[loc_name][i] = 'Waterloo, Ontario, Canada'

In [13]:
data.head(5)

Source Url         Name  \
0  https://www.linkedin.com/profile/view?id=ADEAA...  Nick Begley   
1  https://www.linkedin.com/profile/view?id=ADEAA...  Aaron Patel   
2  https://www.linkedin.com/profile/view?id=ADEAA...   Ashun Shah   
3  https://www.linkedin.com/profile/view?id=ADEAA...     Nancy Li   
4  https://www.linkedin.com/profile/view?id=ADEAA...     Wei Meng   

                    Location           Industry              University  \
0  Waterloo, Ontario, Canada  Computer Software  University of Waterloo   
1     San Francisco Bay Area  Computer Software  University of Waterloo   
2                     Canada  Computer Software     Lakehead University   
3   Windsor, Ontario, Canada  Computer Software  University of Waterloo   
4   Toronto, Ontario, Canada  Computer Software  University of Waterloo   

                  Field Of Study Number Of Connections  \
0           Computer Engineering                   122   
1  Computer Software Engineering                   203   
2               Computer Science                   373   
3               Computer Science                    89   
4     Systems Design Engineering                   202   

                              Job Title           Employer  \
0                     Software Engineer             Google   
1             Software Engineering Coop      WhatsApp Inc.   
2                           QA Engineer  The Working Group   
3  Software Development Engineer Intern      Audible, Inc.   
4             Junior Software Developer    QuickPlay Media   

                           Employment Date    ...    Employment Date 41  \
0  March 2014 – Present (1 year 10 months)    ...                   NaN   
1      September 2015 – Present (4 months)    ...                   NaN   
2        December 2015 – Present (1 month)    ...                   NaN   
3   August 2013 – December 2013 (5 months)    ...                   NaN   
4    July 2014 – Present (1 year 6 months)    ...                   NaN   

  Location 42 University 7 Field Of Study 7 University 8 Field Of Study 8  \
0         NaN          NaN              NaN          NaN              NaN   
1         NaN          NaN              NaN          NaN              NaN   
2         NaN          NaN              NaN          NaN              NaN   
3         NaN          NaN              NaN          NaN              NaN   
4         NaN          NaN              NaN          NaN              NaN   

  University 9 Field Of Study 9 University 10 File name  
0          NaN              NaN           NaN    10.xls  
1          NaN              NaN           NaN    10.xls  
2          NaN              NaN           NaN    10.xls  
3          NaN              NaN           NaN    10.xls  
4          NaN              NaN           NaN    10.xls  

[5 rows x 215 columns]

Write the gigantic data frame to a csv file so that it can be used in the future easily. Name: linkedin_all.csv (done after all the processing).

## Standardize Blackberry
Replace *Research in Motion*, *RIM*, *Blackberry* with RIM. We need to select first the suitable columns to do the replace on.

We will explore the data first to see how often we have relevant cases.

In [14]:
employer_cols = ['Employer'] + ['Employer ' + str(x + 2) for x in range(42)]
nomatchEmp = []
matchesEmp = []
for empl in employer_cols:
    for i in range(len(data[empl])):
        if str(data[empl][i]).lower() == 'blackberry':
            matchesEmp.append(str(data[empl][i]))
        elif str(data[empl][i]).lower() == 'rim':
            matchesEmp.append(str(data[empl][i]))
        elif str(data[empl][i]).lower() == 'research in motion':
            matchesEmp.append(str(data[empl][i]))
        else:
            nomatchEmp.append(str(data[empl][i]))

In [15]:
print('Only Blackberry and related:')
print(pd.Series(matchesEmp).value_counts()[:10])
print('\nAll other employers:')
print(pd.Series(nomatchEmp).value_counts()[:20])

Only Blackberry and related:
BlackBerry            3908
Research In Motion    3440
RIM                    317
Research in Motion      88
Blackberry              66
Research in motion       2
research in motion       2
RESEARCH IN MOTION       1
blackberry               1
rim                      1
dtype: int64

All other employers:
nan                       195613
Microsoft                    459
University of Waterloo       257
Nortel Networks              255
Motorola                     190
Apple                        187
Amazon                       179
IBM                          171
Google                       170
Qualcomm                     158
Ericsson                     147
Nortel                       142
Nokia                        128
Alcatel-Lucent               103
IBM Canada Ltd.               92
QNX Software Systems          70
IBM Canada                    68
NVIDIA                        66
TAT AB                        66
Facebook                      65
dtype:

We will replace the names Research in Motion, RIM, Blackberry with RIM now.

In [16]:
employer_cols = ['Employer'] + ['Employer ' + str(x + 2) for x in range(42)]
for empl in employer_cols:
    for i in range(len(data[empl])):
        if str(data[empl][i]).lower() == 'blackberry':
            data[empl][i] = 'RIM'
        elif str(data[empl][i]).lower() == 'rim':
            data[empl][i] = 'RIM'
        elif str(data[empl][i]).lower() == 'research in motion':
            data[empl][i] = 'RIM'

## Clean the dates
• Step 5 - clean employment date columns
◦ delineate column based on '('
◦ There’s currently a problem in the scrapper so that about 1% of the time this is a very long cell with lots of extraneous info. Right now have to remove them manually else they’ll mess up the text to column function in excel
◦ create new column after employment date column, split cell based on ‘(‘, delate data in new column
◦ next step: need to make this computer readable. Ideas: change all ‘Present’ to 2016, remove all text from columns, allowing us to get years they were employed.

Solve the problem with a lot of extraneous info. The solution is very simple - the date is everything before symbol ';', so check if there is ';' in the cell, and if it is there, keep only the parts before ';'.

We need to know first all the parts which have dates.

In [17]:
date_cols = ['Employment Date'] + ['Employment Date ' + str(x + 2) for x in range(40)] + ['Degree Time'] + ['Degree Time ' + str(x + 2) for x in range(10)]

In [18]:
for date_col in date_cols:
    for i in range(len(data[date_col])):
        pos = str(data[date_col][i]).find(';')
        if pos != -1:
            data[date_col][i] = data[date_col][i][:pos]

Check for any remaining problems which will need to be fixed just manually:

In [19]:
for date_col in date_cols:
    for i in range(len(data[date_col])):
        if len(str(data[date_col][i])) > 50:
            print(data[date_col][i], '\nIndex: ', i, ' Date column:', date_col)

September 2004 – August 2006 (2 years) (Open) 1 honor or award 
Index:  1145  Date column: Employment Date 4
2003 – 2006 (3 years) (Open) 8 recommendations, including 
Index:  4627  Date column: Employment Date 4
2001 – 2001 (less than a year) (Open) 1 recommendation 
Index:  3677  Date column: Employment Date 5


In [20]:
data['Employment Date 4'][1145] = 'September 2004 – August 2006 (2 years)'
data['Employment Date 4'][4627] = '2003 – 2006 (3 years)'
data['Employment Date 5'][3677] = '2001 – 2001 (less than a year)'

Now we want to delete the duration described inside the brackets. If needed, it can be calculated easily from the years given.

In [21]:
for date_col in date_cols:
    for i in range(len(data[date_col])):
        pos = str(data[date_col][i]).find('(')
        if pos != -1:
            data[date_col][i] = data[date_col][i][:pos].strip()

Change Present to November 2016 in all dates.

In [22]:
for date_col in date_cols:
    for i in range(len(data[date_col])):
        data[date_col][i] = str(data[date_col][i]).replace('Present', 'November 2016')

Change the dates so that they can be easily processed by computer. They must be in some standardised format.

Our standardised format will be the following:

MM/YY-MM/YY or YY-YY or YY-MM/YY or MM/YY-YY

If there is some case when there was just one date in a cell, we repeat the date to keep the format (e.g. 2016 --> 2016-2016).

In [23]:
months_num = {'January': '1', 'February': '2', 'March': '3', 'April': '4', 'May': '5', 'June': '6', 'July': '7', 'August': '8', 'September': '9', 'October': '10', 'November': '11', 'December': '12'}
for date_col in date_cols:
    for i in range(len(data[date_col])):
        if str(data[date_col][i]) != 'nan':
            str_date = ''
            parts = data[date_col][i].split('\x96')
            if len(parts) == 2:
                p1 = parts[0].split()
                if len(p1) == 2:
                    str_date = months_num[p1[0]] + '/' + p1[1]
                elif len(p1) == 1:
                    str_date = p1[0]
                p2 = parts[1].split()
                str_date += '-'
                if len(p2) == 2:
                    str_date += months_num[p2[0]] + '/' + p2[1]
                elif len(p2) == 1:
                    str_date += p2[0]
                data[date_col][i] = str_date
            elif len(parts) == 1:
                p1 = parts[0].split()
                if len(p1) == 2:
                    str_date = months_num[p1[0]] + '/' + p1[1] + '-' + months_num[p1[0]] + '/' + p1[1]
                elif len(p1) == 1:
                    str_date = str(int(float(p1[0]))) + '-' + str(int(float(p1[0])))
                data[date_col][i] = str_date
            else:
                data[date_col][i] = np.nan
        else:
            data[date_col][i] = np.nan

In [24]:
data.describe()

Source Url             Name  \
count                                                5291             5291   
unique                                               5291             4790   
top     https://www.linkedin.com/profile/view?id=ADEAA...  LinkedIn Member   
freq                                                    1              307   

                         Location           Industry              University  \
count                        5291               5291                    4666   
unique                        491                 10                     883   
top     Waterloo, Ontario, Canada  Computer Software  University of Waterloo   
freq                          888               2440                     894   

          Field Of Study Number Of Connections          Job Title Employer  \
count               4311                  5290               4981     4981   
unique              1057                   510               2554     2131   
top     Computer Science                  500+  Software Engineer      RIM   
freq                1025                  1037                447      483   

       Employment Date    ...    Employment Date 41   Location 42  \
count             4477    ...                     1             1   
unique             631    ...                     1             1   
top     6/2015-11/2016    ...        7/1990-12/1994  Osaka, Japan   
freq               130    ...                     1             1   

                 University 7 Field Of Study 7       University 8  \
count                      49                3                  8   
unique                     42                3                  7   
top     University of Toronto   Double Honours  McGill University   
freq                        4                1                  2   

           Field Of Study 8                                University 9  \
count                     2                                          15   
unique                    2                                          15   
top     Computer Programmer  University of Illinois at Urbana-Champaign   
freq                      1                                           1   

       Field Of Study 9           University 10 File name  
count                 2                       3      5291  
unique                2                       3        26  
top            Accounts  University of Waterloo    22.xls  
freq                  1                       1       700  

[4 rows x 215 columns]

## Standardise education background
Create schema of education (e.g. BA / Bachelors / BSc / B.Tech) and look for way to standardise.

Idea is to set the standards as Bachelor, Master, Doctorate, Other.

In [25]:
degrees = ['Degree'] + ['Degree ' + str(x + 2) for x in range(10)]
bachelor = set([])
master = set([])
doctorate = set([])
other = set([])
all_degrees = []
for deg in degrees:
    for i in range(len(data[deg])):
        if str(data[deg][i]) != 'nan':
            all_degrees.append(data[deg][i])

In [26]:
print(pd.Series(all_degrees).value_counts())

Bachelor's degree                                                      171
BS                                                                     153
Bachelor of Applied Science (B.A.Sc.)                                  149
Bachelor                                                               142
MS                                                                     134
Bachelor of Engineering (B.Eng.)                                       129
Bachelor of Science                                                    115
Bachelor's Degree                                                      113
Bachelor of Engineering                                                111
BSc                                                                    107
Bachelor of Science (B.Sc.)                                            106
Master's degree                                                        105
Diploma                                                                104
Bachelor of Computer Scie

Heuristics to decide on the type of a degree based on its name - we create a set of the names, but we do it intelligently.

We iterate over the conditions manually, and we improve them to include the items we want there based on what it tells us is there - the conditions are quite clear from the code.

In [27]:
for deg in all_degrees:
    if 'ph.' in deg.lower() or 'phd' in deg.lower() or 'phil' in deg.lower() or 'ph ' in deg.lower() or 'doctorate' in deg.lower():
        doctorate.add(deg)
    elif 'mast' in deg.lower() or 'master' in deg.lower() or 'MS' in deg or 'M ' in deg or 'MA' in deg or 'meng' in deg.lower() or 'M.' in deg or 'mba' in deg.lower() or 'mmath' in deg.lower() or  'minf' in deg.lower() or 'mphys' in deg.lower() or'postgrad' in deg.lower():
        master.add(deg)
    elif 'bach' in deg.lower() or 'bachelor' in deg.lower() or 'beng' in deg.lower() or 'bsc' in deg.lower() or 'bmath' in deg.lower() or 'B ' in deg or 'BA' in deg or 'b.' in deg.lower() or 'ba.' in deg.lower() or 'bs' in deg.lower() or 'bc' in deg.lower() or 'undergrad' in deg.lower():
        bachelor.add(deg)
    else:
        other.add(deg)

We want to rename all of the qualifications so that we can work more easily with them.

In [28]:
for deg in degrees:
    for i in range(len(data[deg])):
        if str(data[deg][i]) != 'nan':
            if str(data[deg][i]) in doctorate:
                data[deg][i] = 'Doctorate'
            elif str(data[deg][i]) in master:
                data[deg][i] = 'Master'
            elif str(data[deg][i]) in bachelor:
                data[deg][i] = 'Bachelor'
            elif str(data[deg][i]) in other:
                data[deg][i] = 'Other'
            else:
                data[deg][i] = np.nan

In [29]:
data.to_csv('linkedin_all.csv', index=False)

## Find way to systematically identify people who have worked in Waterloo AND worked for RIM

'Waterloo, Ontario, Canada' is the location, and 'RIM' is the employer. This should be for a corresponding time period, so the people should have worked in Waterloo and at the same time for RIM (not at different times). So we get only people who have worked at RIM while being in Waterloo.

Employer N, Location N correspond to the same job at given time (also checked manually on LinkedIn). These should be Employer N = 'RIM' Location N = 'Waterloo, Ontario, Canada' for at least one N so that we can accept such person.

Create a separate csv file with only these people.

In [30]:
rows = []
loc_empl = [('Location', 'Employer')] + [('Location ' + str(x + 2), 'Employer ' + str(x + 2))  for x in range(41)]
for i in range(len(data)):
    for e in loc_empl:
        if data.iloc[i][e[0]] == 'Waterloo, Ontario, Canada' and data.iloc[i][e[1]] == 'RIM':
            rows.append(i)
            break
data_rim_w = data.iloc[rows,:]

In [31]:
print('Number of suitable people: ', len(rows))

Number of suitable people:  1115


We have 1115 people who worked at RIM while being in Waterloo! This is quite a good sample (about 1/5 of all people from the files are suitable for the research).

In [32]:
data_rim_w.to_csv('linkedin_rim_waterloo.csv', index=False)

## Identify point where they stopped working for RIM pre- and post- 2008 and extract their subsequent jobs and locations
Since still quite many of them are still working for RIM, we will use 3 categories - present ('finished' working in 11/2016 = present), pre-2008 (< 2009), post-2008 (> 2008). We will include a new column for this named 'Stopped Working'. We are interested only in the latest employment with RIM for a given person. In some cases it may happen that they did not give the dates for the employment, so we leave it blank in such case.

As for extracting the subsequent jobs and locations, we keep them as a list of tuples for each person, the list contains details about each of the subsequent jobs and locations (the latest jobs are the first), and each tuple has the following structure: ('Location', 'Employer', 'Job Title', 'Employment Date'). These are stored directly as a list inside the dataframe - it should be easy to get the meaning from there, and it is very easy to work with it directly here.

In [33]:
data_rim_w = pd.read_csv('linkedin_rim_waterloo.csv', encoding = "ISO-8859-1")

In [34]:
# we will include only employers until employer 41 because we have the dates only until Employment Date 41
stopped_working = []
loc_empl_job_date = [('Location', 'Employer', 'Job Title', 'Employment Date')] + [('Location ' + str(x + 2), 'Employer ' + str(x + 2), 'Job Title ' + str(x + 2), 'Employment Date ' + str(x + 2))  for x in range(40)]
subseq_jobs_loc = []
for i in range(len(data_rim_w)):
    jobs_loc = []
    for e in loc_empl_job_date:
        if data_rim_w.iloc[i][e[0]] == 'Waterloo, Ontario, Canada' and data_rim_w.iloc[i][e[1]] == 'RIM':
            if str(data_rim_w.iloc[i][e[3]]) != 'nan':
                if data_rim_w.iloc[i][e[3]].split('-')[1] == '11/2016':
                    stopped_working.append('present')
                elif int(data_rim_w.iloc[i][e[3]].split('-')[1][-4:]) > 2008:
                    stopped_working.append('post-2008')
                else:
                    stopped_working.append('pre-2008')
            else:
                stopped_working.append(np.nan)
            break
        jobs_loc.append((data_rim_w.iloc[i][e[0]], data_rim_w.iloc[i][e[1]], data_rim_w.iloc[i][e[2]], data_rim_w.iloc[i][e[3]]))
    subseq_jobs_loc.append(jobs_loc)

In [35]:
len(subseq_jobs_loc)

1115

In [42]:
subseq_jobs_loc[:3]

[[('San Francisco Bay Area',
   'WhatsApp Inc.',
   'Software Engineering Coop',
   '9/2015-11/2016'),
  ('Mountain View, CA',
   'Amazon Web Services',
   'Software Engineering Intern',
   '5/2014-8/2014'),
  ('Cambridge, MA', 'Oracle', 'Software Developer', '9/2013-12/2013')],
 [('Vancouver, British Columbia, Canada',
   'Audible, Inc.',
   'Software Development Engineer Intern',
   '8/2015-11/2016'),
  ('United States',
   'SAP',
   'Software Development Engineer Intern',
   '1/2014-12/2014'),
  ('Waterloo, Ontario, Canada',
   'D2L',
   'Software Development Engineer Intern',
   '5/2013-8/2013')],
 [('Waterloo, Ontario, Canada',
   'IBM Canada',
   'Compiler Optimization - Software Developer',
   '9/2015-11/2016'),
  ('Markham, Ontario', 'IBM Canada', 'Software Developer', '9/2014-12/2014'),
  ('Ottawa, Canada Area', 'TELUS', 'Software Developer', '1/2014-4/2014'),
  ('Toronto, ON',
   'Informatica',
   'Software Engineering Intern',
   '5/2013-8/2013'),
  ('Toronto, ON', 'RIM', 'W

In [37]:
pd.Series(stopped_working).value_counts()

post-2008    633
pre-2008     164
present      134
dtype: int64

In [38]:
data_rim_w['Stopped Working'] = stopped_working
data_rim_w['Subsequent Jobs and Locations'] = subseq_jobs_loc

In [39]:
data_rim_w.head()

Source Url            Name  \
0  https://www.linkedin.com/profile/view?id=ADEAA...     Aaron Patel   
1  https://www.linkedin.com/profile/view?id=ADEAA...        Jimmy Fu   
2  https://www.linkedin.com/profile/view?id=ADEAA...      Saad Zaman   
3  https://www.linkedin.com/profile/view?id=ADEAA...     Whitney Mak   
4  https://www.linkedin.com/profile/view?id=ADEAA...  Fikayo Odunayo   

                              Location           Industry  \
0               San Francisco Bay Area  Computer Software   
1  Vancouver, British Columbia, Canada  Computer Software   
2            Waterloo, Ontario, Canada  Computer Software   
3       Richmond Hill, Ontario, Canada  Computer Software   
4            Hamilton, Ontario, Canada  Computer Software   

               University                                     Field Of Study  \
0  University of Waterloo                      Computer Software Engineering   
1  University of Waterloo                               Computer Engineering   
2                     NaN                       Honours Computer Engineering   
3  University of Waterloo  Honours Computer Engineering with Management S...   
4  University of Waterloo                               Computer Engineering   

  Number Of Connections                                   Job Title  \
0                   203                   Software Engineering Coop   
1                  500+        Software Development Engineer Intern   
2                  500+  Compiler Optimization - Software Developer   
3                   147                           Software Engineer   
4                   259                          Software Developer   

        Employer Employment Date  \
0  WhatsApp Inc.  9/2015-11/2016   
1  Audible, Inc.  8/2015-11/2016   
2     IBM Canada  9/2015-11/2016   
3  Zynga Toronto  9/2011-12/2011   
4      Microsoft   1/2014-4/2014   

                         ...                         University 7  \
0                        ...                                  NaN   
1                        ...                                  NaN   
2                        ...                                  NaN   
3                        ...                                  NaN   
4                        ...                                  NaN   

  Field Of Study 7 University 8 Field Of Study 8 University 9  \
0              NaN          NaN              NaN          NaN   
1              NaN          NaN              NaN          NaN   
2              NaN          NaN              NaN          NaN   
3              NaN          NaN              NaN          NaN   
4              NaN          NaN              NaN          NaN   

  Field Of Study 9 University 10 File name Stopped Working  \
0              NaN           NaN    10.xls       post-2008   
1              NaN           NaN    10.xls       post-2008   
2              NaN           NaN    10.xls       post-2008   
3              NaN           NaN    10.xls       post-2008   
4              NaN           NaN    10.xls             NaN   

                       Subsequent Jobs and Locations  
0  [(San Francisco Bay Area, WhatsApp Inc., Softw...  
1  [(Vancouver, British Columbia, Canada, Audible...  
2  [(Waterloo, Ontario, Canada, IBM Canada, Compi...  
3  [(Richmond Hill, Ontario, Canada, Zynga Toront...  
4  [(Hamilton, Ontario, Canada, Microsoft, Softwa...  

[5 rows x 217 columns]

In [40]:
data_rim_w.to_csv('linkedin_rim_waterloo.csv', index=False)